Streaming: Rework websocket server initialisation & authentication code

ThisIsMissEm opened the pull request:
This moves us away from the [deprecated verifyClient mechanism](https://github.com/websockets/ws/issues/377#issuecomment-462152231), to using a custom 'upgrade' event handler as recommended by the author of the `ws` module. 

In the future, we should probably move towards using a custom WebSocket subclass to add in additional methods and data directly to the WebSocket instance, by passing the [`WebSocket` option](https://github.com/websockets/ws/blob/master/doc/ws.md#new-websocketserveroptions-callback), which would give a cleaner abstraction for the logger and other data we may want to attach to the websocket, such as `isAlive` — you can see what that'll look like in this branch: https://github.com/mastodon/mastodon/compare/main...ThisIsMissEm:mastodon:experimental/custom-eventsource-and-websocket-classes?expand=1

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -662,9 +690,6 @@ const startServer = async () => {
   const streamFrom = (ids, req, output, attachCloseHandler, destinationType, needsFiltering = false) => {
-    const accountId = req.accountId || req.remoteAddress;
-
-    log.verbose(req.requestId, `Starting stream from ${ids.join(', ')} for ${accountId}`);
```
This log line ended up being duplicated due to changes to the initialisation logging.

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -13,7 +13,6 @@ const { JSDOM } = require('jsdom');
 const log = require('npmlog');
 const pg = require('pg');
 const dbUrlToConfig = require('pg-connection-string').parse;
-const uuid = require('uuid');
```
This is just dropped in this pull request, not in #27828, but linting was complaining.

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -329,29 +383,6 @@ const startServer = async () => {
-   * @param {function(Error=): void} next
-   */
-  const setRequestId = (req, res, next) => {
-    req.requestId = uuid.v4();
```
There are a bunch of typescript warnings due to dropping `requestId` in this pull request, but these are all resolved by the logging changes in #27828

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -182,14 +181,69 @@ const CHANNEL_NAMES = [
 const startServer = async () => {
+  const pgPool = new pg.Pool(pgConfigFromEnv(process.env));
+  const server = http.createServer();
+  const wss = new WebSocket.Server({ noServer: true });
```
The reason for not passing the `server` option of the http Server above is because we want to use a custom upgrade handler.

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -182,14 +181,69 @@ const CHANNEL_NAMES = [
+        'X-Request-Id': request.id.toString(),
+        // Send the error message via header so it can be debugged in developer
+        // tools:
+        'X-Error-Message': err.message
```
This may leak internal information about errors to the client, potentially, but with the logging changes, we can capture this error and log to the error log instead, with the correlation ID. We likely need to limit this to errors with a status only, but I'm not sure if all code paths that result in this being reached create the errors appropriately.

There's more refactoring needed around the promises used to fetch the account from the request and valid the initial channel join request (from the URL), currently this is done in the `on("connection")` handler, when it'd be better to terminate the connection before even initialising the WebSocket.

ClearlyClaire replied:
I think I'd just leave that out for now, we did not previously return them, so this would not be a regression.

ThisIsMissEm replied:
Yeah, I've removed this out but left a TODO; it was very helpful when debugging the websocket connections in browser for issues (e.g., invalidating the access token on the socket)

ClearlyClaire commented on streaming/index.js:
```diff
@@ -182,14 +181,69 @@ const CHANNEL_NAMES = [
+    socket.on('error', onSocketError);
+
+    // Authenticate:
+    accountFromRequest(request).then(( /* resolvedAccount */) => {
+      wss.handleUpgrade(request, socket, head, function done(ws) {
+        // Remove the error handler:
+        socket.removeListener('error', onSocketError);
+
+        // Start the connection:
+        wss.emit('connection', ws, request);
+      });
+    }).catch((err) => {
```
Shouldn't the error handling only be covering the part before the `handleUpgrade`? i.e. just the `accountFromRequest` part.

Also, this is pretty minor, but I'd have the catch function be a separate helper function.

ThisIsMissEm replied:
Yeah, I might switch this to async/await actually

ClearlyClaire replied:
The helper function is less about reuse than seeing the flow at a glance. I think something like `ws`'s `abortHandshake` makes sense.

ThisIsMissEm replied:
I think in future refactoring this would be a good idea, but for now, I'd be inclined to say that this would just add to the chaos of the `index.js` file that already needs significant refactoring to split it up down the road.

perhaps a compromise would be adding a "TODO" comment for future refactorer?

ThisIsMissEm replied:
I've skipped the abortHandshake change, as it'll make landing #27828 much more difficult.

ClearlyClaire commented on streaming/index.js:
```diff
@@ -1331,11 +1350,6 @@ const startServer = async () => {
       // Decrement the metrics for connected clients:
       connectedClients.labels({ type: 'websocket' }).dec();
-
-      // ensure garbage collection:
-      session.socket = null;
-      session.request = null;
-      session.subscriptions = {};
```
Was this useless? Or did some change make it useless? If so, which?

ThisIsMissEm replied:
This was me being overtly cautious, but I think it's fine to drop (it creates a typescript mess otherwise)

ThisIsMissEm replied:
The long term goal is to not even need this Session object, and instead use a set of maps / lists for managing channel membership. Then anything we need should be on either the `Websocket` or `Eventstream` instance, as noted in https://github.com/mastodon/mastodon/compare/main...ThisIsMissEm:mastodon:experimental/custom-eventsource-and-websocket-classes?expand=1

The only reason we currently need the "request" is because the user information is decorated on to that, which I've since started to refactor to just taking a request and returning a promise of user info.

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -182,14 +181,74 @@ const CHANNEL_NAMES = [
+      });
+
+      // Write the HTTP response manually:
+      socket.end(`HTTP/1.1 ${statusCode} ${http.STATUS_CODES[statusCode]}\r\n${Object.keys(headers).map((key) => `${key}: ${headers[key]}`).join('\r\n')}\r\n\r\n`);
```
@ClearlyClaire fwiw, this code is based on the code from the ws library: https://github.com/websockets/ws/blob/master/lib/websocket-server.js#L490

ThisIsMissEm replied:
It's a little unclear what's going on in that code, but basically the `wss.handleUpgrade` function I'm calling is usually bound directly to the `upgrade` event on the server, if there's a server.

ClearlyClaire commented:
Looks fine overall, but I'd prefer keeping the `requestId` stuff for now, especially since it's still used in places…

ThisIsMissEm commented:
> Looks fine overall, but I'd prefer keeping the `requestId` stuff for now, especially since it's still used in places…

I hear you, but there's a reason this was originally all done in the logging change, because these are all strongly related, imo.

ThisIsMissEm commented:
> Looks fine overall, but I'd prefer keeping the `requestId` stuff for now, especially since it's still used in places…

I hear you, but there's a reason this was originally all done in the logging change, because these are all strongly related, imo.

ClearlyClaire commented:
> > Looks fine overall, but I'd prefer keeping the `requestId` stuff for now, especially since it's still used in places…
> 
> I hear you, but there's a reason this was originally all done in the logging change, because these are all strongly related, imo.

I still don't understand why it's part of this commit and not the other one…

ThisIsMissEm commented:
@ClearlyClaire just to clarify, are you saying this won't be merged as long as the request ID changes are in here? And that you want me to spend 30-60 minutes to pick those out just to do them in another already opened pull request that this was split from at Renauld's request?

Because that's going to delay being able to move forwards with this until I have the time & energy to do those additional changes.

ClearlyClaire commented:
I can try and take this PR over if you prefer.

ThisIsMissEm committed:
Streaming: Rework websocket server authentication & initialisation code

This moves us away from the deprecated verifyClient mechanism, towards using a
custom 'upgrade' event handler as recommended by the author of the ws module.

In the future, we could even move towards using a custom WebSocket subclass to
add in additional methods and data directly to the WebSocket instance.

ThisIsMissEm commented:
@ClearlyClaire have finished rebasing this, and also #27828 — hopefully these are both good to go now.
